thrift_rpc/recorder.go

```diff
@@ -385,6 +387,19 @@ func (r *Recorder) Flush() {
 	}
 }
 
+func (r *Recorder) Close() error {
+	r.lock.Lock()
+	closech := r.closech
+	r.closech = nil
+	r.lock.Unlock()
+
+	if closech != nil {
+		close(closech)
+	}
+
+	return nil
```

Comment:
This returns immediately, which means that a caller of `Close() error` does not have any idea whether the reporting loop has actually shipped its last report. This makes the `Close()` method un-useful (and a bit misleading) when someone tries to flush all transient state before, e.g, exiting a process (see #56).

Now if we consider that `Close` should block for transient reports to be shipped, we hit another problem: what if `Close` blocks for too long? Elsewhere I solved this problem by passing a `timeout` value as a configuration in the constructor (`rec.buffer` holds the pending reports):

```go
close(rec.buffer)
done := make(chan struct{})
go func() {
    wg.Wait()
    close(done)
}()
if tr.flushTimeout != 0 {
    select {
    case <-time.After(tr.flushTimeout):
        return errTimeoutFlush
    case <-done:
    }
} else {
    <-done
}
return nil
```

This gives a chance to the `Tracer` to flush its intermittent spans, while giving some control to the application over its odds of having delivered its spans. 

This may sound a bit ridiculous, but this makes all the difference when integrating tracing in CLI applications, which are very short lived. You may wonder why someone would want to trace CLI applications: turns out it's super useful for internal tools, a bit similarly to how it's super useful to single page JS applications.